feat: Add Terra SRPM and appstream page, remove Terra Policy page and move the contents into contributing, and new guidelines page. Also fix some stuff (#196)
FyraLabs/devdocs: pages/terra/SRPM.mdx

   ├─ path: /home/owen/Documents/Projects/devdocs
   └─ licenseFile: /home/owen/Documents/Projects/devdocs/LICENSE
```

#### %npm_license
Expands to: `/usr/bin/license-checker --limitAttributes licenses`.
Above macro, but limits the output to only the license.

Example output:

```
/usr/bin/license-checker --limitAttributes licenses
└─ @fyralabs/devdocs@@0.0.1
   └─ licenses: MIT
```

### PNPM Macros

#### %vendor_pnpm
Use if PNPM is not usable/available from repo

#### %pnpm_common_envvars
[TODO]

#### %__pnpx
[TODO]

#### %pnpm_build
Should be used in place of `pnpm build`.

#### %pnpm_audit
Audit pnpm if package uses bundled Node dependencies. For use in the `%check` section

#### %pnpm_audit_fix
Fix dependencies using PNPM

#### %pnpm_approve_builds
Approve dependency scripts if needed

### Yarn Macros
[TODO]

#### %yarn_common_envvars
[TODO]

#### %__yarn
Expands to: `/usr/bin/env %{yarn_common_envvars} /usr/bin/yarn`.

#### %__yarn_dlx
[TODO]

#### %yarn_build
Should be used in place of `yarn build`.

#### %yarn_audit
Audit yarn if package uses bundled Node dependencies. For use in the `%check` section

## Nim
These macros assist with building Nim projects.

### %nim_prep

> `nimble`.

Reviewer: ```suggestion
Set up build environment with required packages via `nimble setup`.
```

### %nim_build (or %nim_c)
Builds a nim package.
Requires an argument to `src/pkgname`

### %nim_tflags

> [TODO]

Reviewer: ```suggestion
C Compiler flags to Nim.
```

### %nim_lflags

> [TODO]

Reviewer: ```suggestion
Linker flags to Nim.
```

```
-B: bash (but the completion file doesn't have `.bash` extension)
-b: bash
-e: elvish
-f: fish
-z: zsh
```

For specifying the name (the default will be your `%name`) of the shell completion subpackages (if needed):
```
-n pkgname
```

Can also pass `cmd1`, etc. as the commands the completion files are for.

Example using all flags:
```
%pkg_completion -Bbefz -n pkgname cmd1 cmd2 ...
```

A good example spec that uses this macro is [yadm](https://github.com/terrapkg/packages/blob/d861164790b3d77f5ee094d303de2dfc3299ef8b/anda/devs/yadm/yadm.spec#L22).
This spec creates these packages:

```
yadm-x.x.x-x.fcxx.noarch.rpm - contains all core package files.
yadm-bash-completion-x.x.x-x.fcxx.noarch.rpm - contains `/usr/share/bash-completion/completions/yadm`
yadm-fish-completion-x.x.x-x.fcxx.noarch.rpm - contains `/usr/share/fish/vendor_completions.d/yadm.fish`
yadm-zsh-completion-x.x.x-x.fcxx.noarch.rpm - contains `/usr/share/zsh/site-functions/_yadm`
```

### %pkg_devel_files
[TODO]: flesh this out lol
Set the summary & description, then find development `%files`.

```
%package devel
# generates Summary:, %description devel and the file lists
# use -F to omit %files

> # use -n to manually specify name other than `%name-%devel`

Reviewer: ```
# -n <pkgname>: name of the devel package
# -N <mainpkg>: name of the main package
```

`-F` doesn't exist

%pkg_devel_files
```

### %pkg_static_files
Set the summary & description, then find static libraries.

### %pkg_libs_files

> [TODO]

Reviewer: ```suggestion
Set the summary & description, then find dynamic library files (`%_libdir/*.so.*`).
```

If either of the above are the case for your package, you can use `%zig_build_target` with appropriate options in the `%build` or `%install` section depending on how the project is built (as some can be built directly into root by setting the `DESTDIR=%{buildroot}` variable).

Example:

```rpmspec
%build
%{zig_build_target -r fast -c x86_64_v2} \
  -Demit-docs
```

Note that `-c` also works as a fallback option to the target architecture set by RPM if no following argument is supplied. However, this can **_only_** be used when RPM and Zig's architecture formats are the same, otherwise the build will fail.
For projects where dynamic linking of all dependencies is simply not possible (such as the version in the Fedora repos being too new or too old) or that have no dependencies to link, you can pass the `-s` flag. You can (and should) still dynamically link compatible dependencies by using `-fsys=pkgname`.

More information can be found on our [Zig packaging guidelines](../terra/policy#zig).

## Miscellaneous
General macros that do not fit into a specific category.

> ### %__anda_staticfiles

Reviewer: why is this here but the develfiles are like at the top of this file?

Author: Oh whoops I'll fix that

Expands to: `%_anda_srpm_macros_dir/find-staticinfo.sh %buildroot > anda-staticfiles.list`.
Creates anda-staticfiles.list file in `%{buildroot}` to be used by `%pkg_static_files`.

### %evr
used for declaring how a package interacts with others, for acampe, in a `Requires:` tag.
For example, `Requires: %{name} = %evr`
expands to `%{?epoch:%{epoch}:}%{version}-%{release}`.

### %git_clone
Downloads the source code via `git clone`, and passes useful (in this situation) flags such as`--depth 1`, `-q`, `--recursive-submodules`, and `-j${RPM_BUILD_NCPUS}`.
A git URL and a `%version` or `%commit` SHOULD be passed to this macro.
Below is an example of how a spec that tracks version tages from a GitHub repo, and needs to git clone the source could handle this:

```rpmspec
Version:        0.8.0
---
URL:            https://github.com/ethangreen-dev/lovely-injector
---
%prep
%git_clone %{url} v%{version}
```

> Below is an example of how a spec that tracks the latest commit from a GitHub repo, and needs to git clone the source could handle this:

Reviewer 2: Here as well. 
```suggestion
Below is an example of how a spec that tracks the latest commit from a GitHub repo, and needs to Git clone the source could handle this:
```

```rpmspec
%global commit 9417838c91aab6088778089b9a3e8330bca53fbd
---
URL:           https://github.com/ad-oliviero/uwufetch
---
%prep
%git_clone %{url} %{commit}
```

> `%git_clone` should only be used when downloading a compressed source cannot be used (for example, if the build process expects a git environment or the release tarball doesn't contain submodules).

Reviewer: ```suggestion
`%git_clone` SHOULD only be used when the source tarballs provided by upstream does not include all required sources in the repository (e.g., if the build process expects a git environment or the tarball doesn't contain submodules).
```

> `%git_clone` SHOULD only be used when the source tarballs provided by upstream does not include all required sources in the repository (e.g., if the build process expects a git environment or the tarball doesn't contain submodules).

Reviewer 2: Git is a proper noun.
```suggestion
`%git_clone` SHOULD only be used when the source tarballs provided by upstream does not include all required sources in the repository (e.g., if the build process expects a Git environment or the tarball doesn't contain submodules).
```

- `-U` MAY be used to set either `%u` or `%U` in the desktop file. If using this you MUST NOT use `u` as a macro name for any reason.
  - `%u` means that a desktop file can only read one URL at a time, `%U` means that it can read multiple at a time. These are what are known as [exec keys](https://specifications.freedesktop.org/desktop-entry/latest/exec-variables.html).
  - Defer to what the desktop file originally contained if you are unsure which to use.

Example usage:
- The command `%electron_install -b app -i app -s app -S alsoapp -D -Owayland -U %u -E WaylandWindowDecorations` would install the program with:
  - Icons named `app.png`.
  - A symlink named `app`.
  - A second symlink called `alsoapp`.
  - A desktop file with the `Exec` line reading `Exec=/usr/lib(64)/pkgname/app --enable-features=WaylandWindowDecorations --ozone-platform=wayland %u`.

## Nim
These macros assist with building Nim projects.

### %nim_prep
Set up build environment with required packages via `nimble setup`.

### %nim_build (or %nim_c)
Builds a nim package.
Requires an argument to `src/pkgname`

### %nim_tflags
C Compiler flags to Nim.

### %nim_lflags

> Linker flags to Nim.

Reviewer 2: Proper noun, grammar.
```suggestion
### %nim_build (or %nim_c)
Builds a Nim package.
Requires an argument to `src/pkgname`

### %nim_tflags
C Compiler flags for Nim.

### %nim_lflags
Linker flags for Nim.
```

- `-U` MAY be used to set either `%u` or `%U` in the desktop file. If using this you MUST NOT use `u` as a macro name for any reason.
  - `%u` means that a desktop file can only read one URL at a time, `%U` means that it can read multiple at a time. These are what are known as [exec keys](https://specifications.freedesktop.org/desktop-entry/latest/exec-variables.html).
  - Defer to what the desktop file originally contained if you are unsure which to use.

Example usage:
- The command `%electron_install -b app -i app -s app -S alsoapp -D -Owayland -U %u -E WaylandWindowDecorations` would install the program with:
  - Icons named `app.png`.
  - A symlink named `app`.
  - A second symlink called `alsoapp`.
  - A desktop file with the `Exec` line reading `Exec=/usr/lib(64)/pkgname/app --enable-features=WaylandWindowDecorations --ozone-platform=wayland %u`.

## Nim
These macros assist with building Nim projects.

### %nim_prep
Set up build environment with required packages via `nimble setup`.

### %nim_build (or %nim_c)
Builds a Nim package.
Requires an argument to `src/pkgname`.

### %nim_tflags

> C Compiler flags for Nim.

Reviewer 2: Oops I missed this last time but this is not a proper noun.
```suggestion
C compiler flags for Nim.
```

For example:

```bash
# Lets say package 'foo' has a 16x16 and 64x64 .png
install -Dm644 icons/16x16.png %{buildroot}%{_hicolordir}/16x16/apps/%{name}.svg
install -Dm644 icons/64x64.png %{buildroot}%{_hicolordir}/64x64/apps/%{name}.svg
```

If the file is a `.svg`, use the [`%_scalableiconsdir{:rpmspec}`](#_scalableiconsdir).

### %_scalableiconsdir
Expands to: `%_hicolordir/scalable/apps{:rpmspec}` or `/usr/share/icons/hicolor/scalable/apps{:rpmspec}`.
The directory that SHOULD be used when installing application `.svg` files to.
You do not need to specify any folders for this macro, the file can be added to the end of this macro.
For example:

```bash
install -Dm644 icons/%{name}.svg %{buildroot}%{_scalableiconsdir}/%{name}.svg
```

If the file is a `.png`, use the [`%_hicolordir{:rpmspec}`](#_hicolordir) detailed above.

### %rpmbuilddir
This macro exists because Terra supports EL distributions (Red Hat Enterprise Linux, Alma Linux, etc.) which currently have an RPM version from before the `%{builddir}{:rpmspec}` change to address issues [like this](https://github.com/rpm-software-management/rpm/issues/3151).

> One, albeit clunky, solution is to define `%{builddir}` manually as `%{_builddir}` in specs targeting distros with different versions of RPM where one has a version older than 4.20. We have this macro instead to make the process easier without overriding any builtin RPM macros. For more information on this solution, see [these](https://github.com/rpm-software-management/rpm/issues/3151#issuecomment-2154550735) [comments](https://github.com/rpm-software-management/rpm/issues/3151#issuecomment-2154639192) on the issue linked above.

Reviewer 2: ```suggestion
One, albeit clunky, solution is to define `%{builddir}{:rpmspec}` manually as `%{_builddir}{:rpmspec}` in specs targeting distros with different versions of RPM where one has a version older than 4.20. We have this macro instead to make the process easier without overriding any builtin RPM macros. For more information on this solution, see [these](https://github.com/rpm-software-management/rpm/issues/3151#issuecomment-2154550735) [comments](https://github.com/rpm-software-management/rpm/issues/3151#issuecomment-2154639192) on the issue linked above.
```

Expands to: `%{?builddir}%{?!builddir:%{_builddir}}{:rpmspec}`

## Go
These macros assist with building Go projects.

### %goprep_online
MAY be used to prep a Go package for online builds. Replaces `%goprep{:rpmspec}` in `%prep{:rpmspec}`.

## JavaScript Runtimes
These are macros to assist with the general packaging of applications that use JavaScript runtimes, such as Node.js or Bun.

### Node.js Macros
Macros for using [Node.js](https://nodejs.org). These require use of the [Fedora Node.js macros](https://docs.fedoraproject.org/en-US/packaging-guidelines/Node.js/#_macros) in the `nodejs-packaging` package.

You MUST either set `%global npm_name pkgname{:rpmspec}` or use `%npm_prep -n pkgname{:rpmspec}` when packaging Node modules directly. This MUST be defined as the canonical name of the module as listed on the NPM Registry.

#### %_nvm_dir
The location in the `%{rpmbuilddir}{:rpmspec}` what will be passed to `$NVM_DIR{:rpmspec}`.

Expands to: `%{rpmbuilddir}/.nvm{:rpmspec}`

#### %__nvm

> Calls NVM (Node Version Manager) while setting `$NVM_DIR{:rpmspec}`. Only used if vendoring Node.js.

Reviewer 2: This is an env var so this is probably better?
```suggestion
Calls NVM (Node Version Manager) while setting `$NVM_DIR{:bash}`. Only used if vendoring Node.js.
```

Expands to: `/usr/bin/env NVM_DIR=%{_nvm_dir} /usr/bin/nvm{:rpmspec}`

#### %vendor_nodejs
this maco SHOULD only be used if Node.js packages available from the repos are not usable for some reason.

> You MUST add `nvm` as a build dependancy: `BuildRequires: nvm{:rpmspec}`.

Reviewer 2: Funny typo. fdhdfhjssdfhjfsdhjk
```suggestion
You MUST add `nvm` as a build dependency: `BuildRequires: nvm{:rpmspec}`.
```

What these do:
- Forces all configuration files into the `%{rpmbuilddir}{:rpmspec}` so they are easily locatable and to prevent attempting to write to the host file system for global config.
- Sets the NPM log level to errors (previously `warn` but this included warnings irrelevant to builds or Node modules).
- Disables funding notifications as this does not make sense to display in an automated build and causes log spam.
- Disables update notifications since we are using a packaged version of NPM unless using `%vendor_nodejs{:rpmspec}`.
- Sets color to `always` for more easily readable logs.

#### %npm_buildflags
Explicit NPM configuration flags of the environment variables set by the Macros.

Expands to: `--userconfig=%{rpmbuilddir}/.npmrc --globalconfig=%{rpmbuilddir}/npmrc --cache=%{rpmbuilddir}/.npm --loglevel=error --fund=false --update-notifier=false --color=always --init-module=%{rpmbuilddir}/.npm-init.js`

#### %__npm
Macro to call NPM with predetermined enviroment variables.

Expands to: `/usr/bin/env %{npm_common_envvars} %{?with_vendored_nodejs:%{__nvm} exec npm --}%{!?with_vendored_nodejs:/usr/bin/npm}`

#### %__npx
Remotely run Node modules with NPX.

> Expands to: `/usr/bin/env %{npm_common_envvars} %{?with_vendored_nodejs:%{__nvm} exec npx --}%{!?with_vendored_nodejs:/usr/bin/npx}`

Reviewer 2: ```suggestion
Expands to: `--userconfig=%{rpmbuilddir}/.npmrc --globalconfig=%{rpmbuilddir}/npmrc --cache=%{rpmbuilddir}/.npm --loglevel=error --fund=false --update-notifier=false --color=always --init-module=%{rpmbuilddir}/.npm-init.js{:rpmspec}`

#### %__npm
Macro to call NPM with predetermined enviroment variables.

Expands to: `/usr/bin/env %{npm_common_envvars} %{?with_vendored_nodejs:%{__nvm} exec npm --}%{!?with_vendored_nodejs:/usr/bin/npm}{:rpmspec}`

#### %__npx
Remotely run Node modules with NPX.

Expands to: `/usr/bin/env %{npm_common_envvars} %{?with_vendored_nodejs:%{__nvm} exec npx --}%{!?with_vendored_nodejs:/usr/bin/npx}{:rpmspec}`
```

#### %bun_pm_trust
Add Node modules to trusted dependencies so their postinstall scripts may be run. Should only be used if necessary.

Available flags:
- `-a{:rpmspec}` can be used to trust all dependencies, otherwise simply list the dependencies you want to trust after the command.

Expands to: `%{__bun} pm trust{:rpmspec}`

### PNPM Macros
Macros for using [PNPM](https://pnpm.io).

#### %_pnpm_home
Where PNPM files and cache are stored.

Expands to: `%{rpmbuilddir}/.pnpm{:rpmspec}`

#### %_pnpm_store
Where Node modules installed by PNPM usually go. This will usually not be produced when using PNPM as a build tool, but exists for rare builds it may be created in.

Expands to: `%{_pnpm_home}/store{:rpmspec}`

#### %pnpm_common_envvars
Sets the NPM evironment variables and the PNPM home dir.

Expands to: `%{npm_common_envvars} PNPM_HOME=%{_pnpm_home}{:rpmspec}`

#### %__pnpm

Expands to: `/usr/bin/env %{pnpm_common_envvars} %{?with_vendored_pnpm:%{rpmbuilddir}/pnpm/pnpm}%{!?with_vendored_pnpm:/usr/bin/pnpm}{:rpmspec}`

> Should be used in place of calling `pnpm` itself.

Reviewer 2: I assume you also meant to do this for the other ones too?
```suggestion
SHOULD be used in place of calling `pnpm` itself.
```

Note that `-c` also works as a fallback option to the target architecture set by RPM if no following argument is supplied. However, this can **_only_** be used when RPM and Zig's architecture formats are the same, otherwise the build will fail.
For projects where dynamic linking of all dependencies is simply not possible (such as the version in the Fedora repos being too new or too old) or that have no dependencies to link, you can pass the `-s` flag. You can (and should) still dynamically link compatible dependencies by using `-fsys=pkgname`.

More information can be found on our [Zig packaging guidelines](../terra/policy#zig).

## Miscellaneous
General macros that do not fit into a specific category.

### %evr
Used for declaring how a package interacts with others, such as, in a `Requires:{:rpmspec}` or `Obsoletes:{:rpmspec}` tag.
For example, `Requires: %{name} = %evr{:rpmspec}`.
Expands to: `%{?epoch:%{epoch}:}%{version}-%{release}`.
Real package examples:

`Requires:` example:

```rpmspec
BuildRequires:  pkgconfig(pkgconf)
BuildRequires:  pkgconfig(zlib)
BuildRequires:  pkgconfig(openssl)
BuildRequires:  pkgconfig(libarchive) >= 3.3.3
Requires:       %name-lib = %evr
```

> `Recommends:` example:

Reviewer 2: ```suggestion
`Recommends:{:rpmspec}` example:
```

```rpmspec
Name:           rnote
License:        GPL-3.0
Recommends:     rnote-cli = %evr
```

> `Provides:` example:

Reviewer 2: ```suggestion
`Provides:{:rpmspec}` example:
```

You may need to refer to the [miscellaneous packaging guidelines](./policy#miscellaneous-packaging-guidelines) for non-macro related Terra guidelines.

If you would like to add/fix/update these macros, or MUST raise an issue, please do so in the [SRPM macros repository](https://github.com/terrapkg/srpm-macros), or reach out to us somewhere [here](https://wiki.ultramarine-linux.org/en/community/community/).

For upstream macros available in Fedora/Terra, check out our unofficial [macros list](../rpm/macros).

<Callout type="info">
  The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL
  NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and
  "OPTIONAL" in this document are to be interpreted as described in
  [RFC 2119](https://www.rfc-editor.org/rfc/rfc2119) and
  [RFC 8174](https://www.rfc-editor.org/rfc/rfc8174).
</Callout>

## Directories
These macros define certain system directories, and SHOULD be used in place of the raw directory path.

### %_appsdir
Expands to: `%_datadir/applications{:rpmspec}` or `/usr/share/applications{:rpmspec}`

> This may seem like a directory to place the application itself into, but this directory is used for applicaiton's `.desktop` files.

Reviewer 2: ```suggestion
This may seem like a directory to place the application itself into, but this directory is used for application's `.desktop` files.
```
